src/coreclr/jit/importer.cpp

```diff
@@ -843,7 +843,32 @@ GenTree* Compiler::impStoreStruct(GenTree*         store,
             // TODO-Bug?: verify if flags matter here
             GenTreeFlags indirFlags = GTF_EMPTY;
             GenTree*     destAddr   = impGetNodeAddr(store, CHECK_SPILL_ALL, &indirFlags);
-            NewCallArg   newArg     = NewCallArg::Primitive(destAddr).WellKnown(wellKnownArgType);
+
+            // Make sure we don't pass something other than a local address to the return buffer arg.
+            // It is allowed to pass current's method return buffer as it is a local too.
+            if (fgAddrCouldBeHeap(destAddr) && !eeIsByrefLike(srcCall->gtRetClsHnd))
+            {
+                unsigned tmp = lvaGrabTemp(false DEBUGARG("stack copy for value returned via return buffer"));
+                lvaSetStruct(tmp, srcCall->gtRetClsHnd, false);
+
+                // Make sure the address is evaluated first if needed.
+                if (store->OperIs(GT_STOREIND, GT_STORE_BLK) &&
+                    ((store->AsIndir()->Addr()->gtFlags & GTF_ALL_EFFECT) != 0))
+                {
+                    unsigned lclNum = lvaGrabTemp(true DEBUGARG("fgMakeTemp is creating a new local variable"));
+                    impStoreToTemp(lclNum, store->AsIndir()->Addr(), curLevel, pAfterStmt, di, block);
```

Comment:
I don't think we can use `impStoreStruct` here since this function is called from outside import via `gtNewTempStore`. I think it needs to create a comma, or only call this function in some cases (see the checks below for `GT_COMMA`)

Reply:
um.. what `impStoreStruct`  are your referring here? did you mean impStoreToTemp? Also, seems like this function already appends stuff to statements so it's weird expect it to not?

Reply:
I guess it's just an implicit contract that when gtNewTempStore calls it - it does not?

Reply:
Yes I meant `impStoreTemp`

> seems like this function already appends stuff to statements so it's weird expect it to not?

Where does it do that? I think we only do that for the `GT_COMMA` case, and it has guards to ensure that only happens during import

Reply:
We spoke offline and Egor convinced me that actually no reordering is happening here, so we don't need to do any spilling here. Sorry about that.